stout/range.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include "stout/stream.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace stout {
+namespace eventuals {
+
+////////////////////////////////////////////////////////////////////////
+
+struct _Range {
+  template <typename K_, typename Arg_>
+  struct Continuation : public detail::TypeErasedStream {
+    // NOTE: explicit constructor because inheriting 'TypeErasedStream'.
+    Continuation(K_ k, int from, int to, int step)
+      : k_(std::move(k)),
+        from_(from),
+        to_(to),
+        step_(step) {}
+
+    void Start() {
+      k_.Start(*this);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    void Ended() {
+      k_.Ended();
+    }
+
+    void Next() override {
+      if (from_ == to_
+          || step_ == 0
+          || (from_ > to_ && step_ > 0)
+          || (from_ < to_ && step_ < 0)) {
+        k_.Ended();
+        return;
+      }
+
+      int temp = from_;
+      from_ += step_;
+      k_.Body(temp);
```

Comment:
```suggestion
      if (from_ == to_
          || step_ == 0
          || (from_ > to_ && step_ > 0)
          || (from_ < to_ && step_ < 0)) {
        k_.Ended();
      } else {
        int temp = from_;
        from_ += step_;
        k_.Body(temp);
      }
```

Reply:
In general let's avoid using `return;` because having complete `if () {} else {}` blocks makes it harder to introduce control flow bugs when we refactor!

Reply:
Hahaha, I'd asked Zakhar to do the exact opposite here previously. 😆 

I like using `return` because it avoids indentation. In general, I like structuring code as:
```
if (something is not normal) {
  handle it;
  return;
}
if (something else is off) {
  handle it;
  return;
}
handle normal case;
with more code;
than the abortive cases;
```
As opposed to:
```
if (something is normal) {
  if (something else is not off) {
    handle normal case;
    with more code;
    than the abortive cases;
  } else {
    handle it;
  }
} else {
  handle it;
}
```
I find the upper version to be much more readable.

Thoughts?

Reply:
1 more vote for using `if else` from me